sharktank/sharktank/layers/mixture_of_experts_block.py

```diff
@@ -60,41 +61,53 @@ def __init__(
                 )
         self.expert_used_count = expert_used_count
         self.expert_count = expert_count
+        self.expert_shared_count = expert_shared_count
         self.n_expert_groups = n_expert_groups
         self.n_limited_groups = n_limited_groups
         self.score_experts = score_experts
         self.normalize_experts = normalize_experts
-        self.add_residual = add_residual
-
-        # Add router gate
-        self.add_module("ffn_gate_inp", LinearLayer(theta("ffn_gate_inp")))
+        self.route_scale = route_scale
+        self.shard_count = shard_count
 
-        self.ffn_norm = torch.nn.Identity()
         self.layer_output_norm = torch.nn.Identity()
-        self.shared_experts = None
-        self.route_scale = None
-        if route_scale is not None and route_scale != 1:
-            self.route_scale = route_scale
+        self.ffn_gate_inp = torch.nn.Identity()
+        self.name_map = {
+            # Experts name mapping for FFN block
+            "ffn_gate_shexp": "ffn_gate",
+            "ffn_up_shexp": "ffn_up",
+            "ffn_down_shexp": "ffn_down",
+            "ffn_gate_exps": "ffn_gate",
+            "ffn_up_exps": "ffn_up",
+            "ffn_down_exps": "ffn_down",
+        }
 
-        # Add FFN norm
-        if "ffn_norm" in theta:
-            self.ffn_norm = RMSNormLayer(theta("ffn_norm"), epsilon=rms_epsilon)
+        # Add router gate
+        if theta.optional_tensor("ffn_gate_inp") is not None:
+            self.add_module("ffn_gate_inp", LinearLayer(theta("ffn_gate_inp")))
 
         # Add expert_count x FFN
         if isinstance(experts_ffn_moe_block, str):
             if experts_ffn_moe_block == "PreGatherFFNMOE":
-                self.experts = PreGatherFFNMOE(theta, activation=moe_activation)
+                self.routed_experts = PreGatherFFNMOE(
+                    theta, activation_fn=moe_activation
+                )
             elif experts_ffn_moe_block == "DenseFFNMOE":
-                self.experts = DenseFFNMOE(theta, activation_fn=moe_activation)
+                ffn_theta = theta.pop("exps", inplace=True)
+                ffn_theta = ffn_theta.rename_tensors(self.name_map)
+                self.routed_experts = DenseFFNMOE(
+                    ffn_theta, expert_count=expert_count, activation_fn=moe_activation
+                )
             else:
                 raise ValueError(
                     f'Unknown experts_ffn_moe_block "{experts_ffn_moe_block}"'
                 )
         else:
-            self.experts = experts_ffn_moe_block
+            self.routed_experts = experts_ffn_moe_block
 
-        if "shared_experts" in theta:
-            self.shared_experts = FFN(theta("shared_experts"))
+        if self.expert_shared_count is not None:
+            ffn_theta = theta.pop("shexp", inplace=True)
```

Comment:
It would be quite surprising for a caller to realize that the theta was modified.